src/libOpenImageIO/imagebufalgo_xform.cpp

```diff
@@ -1144,4 +1144,261 @@ ImageBufAlgo::rotate(const ImageBuf& src, float angle, string_view filtername,
+                    out_iter[chan] = sample_accum[idx] / total_weight;
+                }
+            } else {
+                for (int chan = roi.chbegin; chan < roi.chend; ++chan) {
+                    out_iter[chan] = 0;
+                }
+            }
+            ++out_iter;
+        }
+    });  // end of parallel_image
+    return true;
+}
+
+
+
+static bool
+check_st_warp_args(ImageBuf& dst, const ImageBuf& src, const ImageBuf& stbuf,
+                   int chan_s, int chan_t, ROI& roi)
+{
+    // Validate ST buffer
+    if (!stbuf.initialized()) {
+        dst.error("ImageBufAlgo::st_warp : Uninitialized ST buffer");
+        return false;
+    }
+
+    const ImageSpec& stSpec(stbuf.spec());
+    // XXX: Wanted to use `uint32_t` for channel indices, but I don't want to
+    // break from the rest of the API and introduce a bunch of compile warnings.
+    if (chan_s >= stSpec.nchannels) {
+        dst.errorfmt("ImageBufAlgo::st_warp : Out-of-range S channel index: {}",
+                     chan_s);
+        return false;
+    }
+    if (chan_t >= stSpec.nchannels) {
+        dst.errorfmt("ImageBufAlgo::st_warp : Out-of-range T channel index: {}",
+                     chan_t);
+        return false;
+    }
+    // N.B. We currently require floating-point ST channels
+    if (!stSpec.format.is_floating_point()) {
+        dst.error("ImageBufAlgo::st_warp : ST buffer must be holding floating-"
+                  "point data");
+        return false;
+    }
+
+    if (dst.initialized()) {
+        // XXX: Currently requiring the pixel scales of ST and a preallocated
+        // output buffer to match.
+        if (dst.spec().full_width != stSpec.full_width
+            || dst.spec().full_height != stSpec.full_height) {
+            dst.error("ImageBufAlgo::st_warp : Output and ST buffers must have "
+                      "the same full width and height");
+            return false;
+        }
+    }
+
+    // Prep the dest spec using the channels from `src`, and the intersection of
+    // `roi` and the ROI from the ST buffer (since the ST warp is only defined
+    // for pixels in the ST buffer). We grab a copy of the input ROI before
+    // `IBAprep`, since we want to intersect it ourselves.
+    ROI inputROI(roi);
+    bool res
+        = ImageBufAlgo::IBAprep(roi, &dst, &src,
+                                ImageBufAlgo::IBAprep_NO_SUPPORT_VOLUME
+                                    | ImageBufAlgo::IBAprep_NO_COPY_ROI_FULL);
+    if (res) {
+        roi         = roi_intersection(inputROI, stSpec.roi());
+        roi.chbegin = std::max(roi.chbegin, 0);
+        roi.chend   = std::min(roi.chend, src.spec().nchannels);
+
+        ImageSpec destSpec(dst.spec());
+        destSpec.set_roi(roi);
+        destSpec.set_roi_full(stSpec.roi_full());
+        // XXX: It would be nice to be able to tell `IBAprep` to skip the buffer
+        // initialization. Worth adding a new prep flag?
+        dst.reset(destSpec);
```

Comment:
I don't love this double-reset behavior, but I was trying to lean on `IBAprep` as much as possible, while still applying my own ROI overrides.

I'd like to improve this in some way, particularly since I think the worst-case scenario is the buffer storage being reallocated three times (since this function gets called twice when using the overloads that take a filter name and width).

Reply:
> If dst is already initialized and we're writing into it, won't this throw that out and reallocate a new blank image?

Yes, it will.

> What are the circumstances in which there's an extra reset here?

`IBAprep` will reset it if it's not already initialized, and then I come along and do it again.

I think this is something I could use your perspective on.

The basic issue is that I need to make sure `dst` is "big" enough to fit the ST buffer, and this was just the blunt-instrument solution.

For an already-initialized `dst` buffer, there are two possibilities:

1. The ST buffer's ROI is a subset of `dst`'s.
    * In this case, we should hypothetically be able to continue on without doing any more sanity-checking, and the actual `st_warp` operation would only fill in the subset of the image area defined by the ST buffer's ROI (if that sounds like reasonable behavior).
    * I think `IBAprep` will still alter the ROI argument to match the destination buffer's, which isn't actually what I want to operate on, so I would still need to modify it after the fact.
2. The ST buffer's ROI is a proper superset of `dst`'s (that is, it contains coordinates that are outside of `dst`'s ROI).
    * Since I don't want the operation to be clamped to `dst`'s display window, I think using `ImageBuf::Iterator` to avoid out-of-bounds writes is off the table here.

Maybe the right thing to do in either case would just be to clamp the loop iterator to the intersection of the `dst` and ST buffer ROIs? I think that would allow me to be a little less heavy-handed in this validation/setup code, but maybe I'm under- or overthinking this?

> Yes, it's totally fine to make a new flag value for IBAprep_NO_RESET_DST or something like that. That's not an external API, we can change it however we need to make implementation of these functions easier.

OK, good to know. For context This idea stems from the behavior mentioned in `1.` above, where `IBAprep` will basically use the `dst` buffer to resize the ROI argument. In the cast of an uninitialized `dst` buffer, I want to set up the ROI of the `dst` buffer myself and only allocate the storage one time, but I'd still like to be able to take advantage of some of what `IBAprep` does in terms of validation, metadata copying, etc.

I will say that I'm not sure how `IBAprep` would go about setting up a spec for `dst` without allocating local storage for it (keeping in mind that I'd like to potentially modify the assigned spec further before allocating local storage). Maybe just using `dst.specmod() = spec;` would be appropriate? If so, what would be the most idiomatic way for the caller to actually allocate the underlying storage later on? `ImageBufAlgo::zero`?

Reply:
I think there are other examples of IBA function that take 2 input images and make one output, and I'm not sure why this shouldn't just use the same logic about how big to make dst.

Reply:
OK thanks, I'll have another look.

One thing I keep coming back to when it comes to the idea of `dst`-centric iteration is that the transform is undefined outside the ROI of `stbuf`. Thus, I really only want to operate within the bounds of the ST ROI no matter what, and I don't want to rely on any kind of auto-clamping/wrapping of the `stbuf` iterator to give me default pixel values, etc.

I think pre-clamping the operating ROI to the intersection of the `dst` and `stbuf` ROIs sounds like it should give me the right balance here, as well as getting rid of this extra `dst` reallocation. I'm working on this approach right now.

Reply:
OK, I defer to your best judgment. Clamp the iteration to the well-defined part of stbuf, and we can always revisit this to make it more flexible if user experience shows that to be an issue.

Reply:
OK, I've simplified this logic and removed the extra buffer reset.